catalog/spatial/csw/spatial-csw-endpoint/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/endpoint/CswEndpoint.java

```diff
@@ -649,27 +724,66 @@ private int deleteRecords(DeleteAction deleteAction)
       throws CswException, FederationException, IngestException, SourceUnavailableException,
           UnsupportedQueryException, InterruptedException, ParseException, CQLException {
 
-    deleteAction = transformDeleteAction(deleteAction);
+    final DeleteAction transformDeleteAction = transformDeleteAction(deleteAction);
 
     QueryRequest queryRequest =
-        queryFactory.getQuery(deleteAction.getConstraint(), deleteAction.getTypeName());
+        queryFactory.getQuery(
+            transformDeleteAction.getConstraint(), transformDeleteAction.getTypeName());
 
     queryRequest =
         queryFactory.updateQueryRequestTags(
             queryRequest,
-            schemaTransformerManager.getTransformerSchemaForId(deleteAction.getTypeName()));
+            schemaTransformerManager.getTransformerSchemaForId(
+                transformDeleteAction.getTypeName()));
 
     int batchCount = 1;
     int deletedCount = 0;
 
     String[] idsToDelete = getNextQueryBatch(queryRequest);
 
+    CompletionService<DeleteResponse> deleteCompletionService =
+        new ExecutorCompletionService<>(queryExecutor);
+
     while (idsToDelete.length > 0) {
-      DeleteRequestImpl deleteRequest = new DeleteRequestImpl(idsToDelete);
       LOGGER.debug(
           "Attempting to delete {} metacards from batch {}.", idsToDelete.length, ++batchCount);
-      DeleteResponse deleteResponse = framework.delete(deleteRequest);
-      deletedCount += deleteResponse.getDeletedMetacards().size();
+      final Subject subject = SecurityUtils.getSubject();
+      for (final String id : idsToDelete) {
+        Callable<DeleteResponse> callable =
+            () -> {
+              try {
+                DeleteRequestImpl deleteRequest = new DeleteRequestImpl(id);
```

Comment:
Is there a reason we don't batch the ids in the DeleteRequest?

Reply:
Things were unbatched in order to facilitate parallel calls to the CatalogFramework, with the goal of supporting clients that may not have the ability to make calls in parallel.

Reply:
Are there any performance concerns when deleting a large amount of records?

Reply:
Sorry. I meant more for if a client will only single thread connections to us and send us batches of insert, update, delete with multiple records. This is to parallel process those. For delete if they send us a list of items to delete, this is to facilitate parallel processing through the catalog framework as our API for delete returns the record being deleted, so there is some work the framework has to do to retrieve the record as well.